src/wifi/model/periodic-dmg-wifi-scheduler.cc

```diff
@@ -0,0 +1,539 @@
+    .SetParent<DmgWifiScheduler> ()
+    .SetGroupName ("Wifi")
+    .AddConstructor<PeriodicDmgWifiScheduler> ();
+  return tid;
+}
+
+PeriodicDmgWifiScheduler::PeriodicDmgWifiScheduler ()
+{
+  NS_LOG_FUNCTION (this);
+}
+
+PeriodicDmgWifiScheduler::~PeriodicDmgWifiScheduler ()
+{
+  NS_LOG_FUNCTION (this);
+}
+
+void
+PeriodicDmgWifiScheduler::DoDispose ()
+{
+  NS_LOG_FUNCTION (this);
+  DmgWifiScheduler::DoDispose ();
+}
+
+void
+PeriodicDmgWifiScheduler::UpdateStartAndRemainingTime (void)
+{
+  // for the periodic scheduler, m_allocationStartTime is useless, since the addition
+  // of new SPs is consecutive 
+  NS_LOG_FUNCTION (this);
+  if (m_addtsAllocationList.empty ())
+    {
+      // no existing allocations
+      m_remainingDtiTime = m_dtiDuration.GetMicroSeconds ();
+      // clear the list of available slots: if no allocations have been scheduled, then the DTI is completely available
+      m_availableSlots.clear ();
+      m_availableSlots.push_back (std::make_pair (0, m_dtiDuration.GetMicroSeconds ()));
+    }
+  else
+    {
+      // if there are existing allocations, update DTI time just for consistency
+      m_remainingDtiTime = 0;
+      for (const auto & slot: m_availableSlots)
+        {
+          m_remainingDtiTime += (slot.second - slot.first + 1);
+        }
+    }
+}
+
+void
+PeriodicDmgWifiScheduler::AdjustExistingAllocations (AllocationFieldListI iter, uint32_t duration, bool isToAdd)
+{
+  NS_LOG_FUNCTION (this << duration << isToAdd);
+
+  // This method is called upon a DelTsRequest or after the cleanup of 
+  // non-pseudostatic allocations.
+  // In this version of the periodic scheduler, existing allocations are not shifted 
+  // to fill the created gaps but only the vector listing the available slots is updated.
+  // For this reason, the current input parameters are useless.
+
+  auto addtsListCopy = m_addtsAllocationList;
+
+  // sort the copy to simplify the process of going through the allocation list
+  sort (addtsListCopy.begin (),
+        addtsListCopy.end (),
+        [](const AllocationField& lhs, const AllocationField& rhs){
+      return lhs.GetAllocationStart () < rhs.GetAllocationStart ();
+    });
+
+  uint32_t startAlloc, endAlloc;
+
+  // clear m_availableSlots and refill it based on the updated m_addtsAllocationList
+  m_availableSlots.clear ();
+  m_availableSlots.push_back (std::make_pair (0, m_dtiDuration.GetMicroSeconds ()));
+
+  for (const auto & allocation: addtsListCopy)
+    {
+      startAlloc = allocation.GetAllocationStart ();
+      endAlloc = startAlloc + allocation.GetAllocationBlockDuration () + m_guardTime;
+      if (allocation.GetNumberOfBlocks () > 1)
+        {
+          // if the number of blocks allocated is > 1, the allocation is periodic and must be dealt with accordingly
+          uint8_t blocks = allocation.GetNumberOfBlocks ();
+          while (blocks > 0)
+            {
+              UpdateAvailableSlots (startAlloc, endAlloc);
```

Comment:
don't you need to update `m_remainingDtiTime`?

Reply:
there are more points in the code where i have this doubt related to `UpdateAvailableSlots`, please look at the previous comment